### PR TITLE
Add 90th and 99th percentile frame rasterize time for timeline_summary

### PR DESCRIPTION
**Pull request by liyuqian**
Our first step towards https://github.com/flutter/flutter/issues/18727

### CODE REVIEW

File: packages/flutter_driver/lib/src/driver/timeline_summary.dart [hunk 1]
```diff
@@ -159,6 +168,19 @@ class TimelineSummary {
...
+    if (durations.isEmpty)
+      return null;
+
+    assert(percentile >= 0 && percentile <= 100);
+
+    List<double> doubles = durations.map<double>((Duration duration) => duration.inMilliseconds.toDouble()).toList();
+    doubles.sort();
+
+    return doubles[((doubles.length - 1) * (percentile / 100)).round()];
+
```
Comment (yjbanov) [inline packages/flutter_driver/lib/src/driver/timeline_summary.dart]: nit: extra blank line

File: packages/flutter_driver/lib/src/driver/timeline_summary.dart [hunk 2]
```diff
@@ -159,6 +168,19 @@ class TimelineSummary {
     return total / durations.length;
   }
 
+  double _percentileInMillis(Iterable<Duration> durations, double percentile) {
+    if (durations.isEmpty)
+      return null;
+
+    assert(percentile >= 0 && percentile <= 100);
```
Comment (yjbanov) [inline packages/flutter_driver/lib/src/driver/timeline_summary.dart]: To be more precise, let's compare against `0.0` and `100.0`. In Dart `0` and `100` are integers, and it's not immediately obvious in which direction type coercions are happening.

**Review by yjbanov — APPROVED** [review]

**Comment by yjbanov** [comment]
lgtm

File: packages/flutter_driver/test/src/timeline_summary_test.dart [hunk 3]
```diff
@@ -31,6 +31,17 @@ void main() {
       'name': 'GPURasterizer::Draw', 'ph': 'E', 'ts': timeStamp
     };
 
+    List<Map<String, dynamic>> rasterizeTimeSequenceInMillis(List<int> sequence) {
+      final List<Map<String, dynamic>> result = <Map<String, dynamic>>[];
+      int t = 0;
+      for(int duration in sequence) {
```
Comment (Hixie) [inline packages/flutter_driver/test/src/timeline_summary_test.dart]: nit: space after `for`

File: packages/flutter_driver/lib/src/driver/timeline_summary.dart [hunk 4]
```diff
@@ -159,6 +168,17 @@ class TimelineSummary {
     return total / durations.length;
   }
 
+  double _percentileInMillis(Iterable<Duration> durations, double percentile) {
+    if (durations.isEmpty)
+      return null;
```
[the remark this answers is not in the document]
Comment (liyuqian) [reply packages/flutter_driver/lib/src/driver/timeline_summary.dart]: The `if empty return null` logic was originally added to `_averageInMillis` and `_maxInMillis` by @yjbanov in https://github.com/flutter/flutter/pull/7190 and further edited by @abarth in https://github.com/flutter/flutter/pull/8113 . If we throw here, maybe we need to also throw in those places for consistency. What's your opinion @yjbanov @abarth?

[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter_driver/lib/src/driver/timeline_summary.dart]: Yeah, I think the test script will fail with a cryptic error message if we return `null` here. Might as well throw proactively with something more descriptive.

**Pull request merged**